### PR TITLE
Fix: Unit tests for TDR plugin should use canned bundle verbatim (#2374)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Author (check before every review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `DCP/1` label                                   <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes          <sub>or this PR does not need porting</sub> 
- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain            <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox`                             <sub>or added `no sandbox` label</sub>
- [x] Reindexed `sandbox`                                   <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [ ] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain                                       <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [x] Commented on demo expectations                        <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Reindexed `dev`                                       <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

File: scripts/convert_bundle_v2.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            assert entry['version'] == bundle.version
+
+
+def main(argv):
+    """
+    Load a canned bundle from DCP/1 and write *.manifest.tdr and *.metadata.tdr
+    files showing the desired output for DCP/2.
+    """
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument('--bundle-uuid', '-b',
```
Comment (danielsotirhos) [inline scripts/convert_bundle_v2.py:116]:
```suggestion
    parser.add_argument('--bundle-uuid', '-b',
                        requried=True,
```

File: scripts/convert_bundle_v2.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                        help='The UUID of the existing DCP/1 canned bundle.')
+    parser.add_argument('--source-uuid', '-s',
```
Comment (danielsotirhos) [inline scripts/convert_bundle_v2.py:118]:
```suggestion
    parser.add_argument('--source-uuid', '-s',
                        default='cafebabe-feed-4bad-dead-beaf8badf00d',
```

File: test/service/test_tdr.py [hunk 3]
```diff
@@ -421,20 +365,33 @@ def _bq_schema(self, columns: Mapping[str, type]) -> JSONs:
...
         ]
 
-    def _create_table(self, dataset_name: str, table_name: str, schema: JSONs, rows: JSONs) -> None:
-        # TinyQuery's errors are typically not helpful in debugging missing/extra columns in the row JSON.
+    def _create_table(self,
+                      dataset_name: str,
+                      table_name: str,
+                      schema: JSONs,
+                      rows: JSONs) -> None:
+        # TinyQuery's errors are typically not helpful in debugging missing/
```
Comment (danielsotirhos) [inline test/service/test_tdr.py:373]:
```suggestion
        # TinyQuery's errors are typically not helpful in debugging missing or
```

**Review by danielsotirhos — APPROVED** [review]
Looks great.
Just a couple nit-pick suggestion to leave or take.

File: scripts/convert_bundle_v2.py (hunk 2 above)
Comment (noah-aviel-dove) [reply scripts/convert_bundle_v2.py:118]: Personally I'd prefer that this also be required, we'll see what Hannes thinks.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
A canned TDR subgraph doesn't need to be in the same format as a canned DSS bundle, it should be in a format that is most suitable for loading into TinyQuery or whatever you use to mock the BQ tables. I am guessing that a JSON tree of the following structure would work:

```json
{
    "tables": {
        "links": {
            "rows": [
                {
                    "columns": {
                        "links_id": "…"
                    }
                }
            ]
        }
    }
}
```

This wasn't exactly spelled out in the ticket, so no bump.

**Review by danielsotirhos — APPROVED** [review]
Nicely done.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Please remove all test support for indexing datasets. We'll be phasing that out.

Looking at the cans, I don't see any resemblance with my [suggestion](https://github.com/DataBiosphere/azul/pull/2411#pullrequestreview-526541302). Please discuss this—and any responses to my comments—on parking lot.

Please check your commit titles for spelling errors. I saw one.

File: src/azul/indexer/__init__.py [hunk 4]
```diff
@@ -58,6 +58,9 @@ def for_fqid(cls, fqid: BundleFQID, *, manifest: MutableJSONs, metadata_files: M
     def fquid(self):
         return BundleFQID(self.uuid, self.version)
 
+    def concrete_entity_type(self, entity_key):
```
[the remark this answers is not in the document]
Comment (noah-aviel-dove) [reply src/azul/indexer/__init__.py:61]: @hannes-ucsc: "s/entity_key/file_name"

File: src/azul/json.py [hunk 5]
```diff
@@ -29,6 +31,20 @@ def copy_json(o: JSON) -> MutableJSON:
     return cast(MutableJSON, copy.deepcopy(o))
 
 
+def copy_jsons(o: JSONs) -> MutableJSONs:
```
[the remark this answers is not in the document]
Comment (noah-aviel-dove) [reply src/azul/json.py:34]: Can't see any way to do it.

**Comment by noah-aviel-dove** [comment]
@hannes-ucsc Unfortunately I failed to segregate my fixup changes from the changes I made while resolving conflicts, so no `fixup` commits. I tried to go back and separate them but it was excessively difficult, time-consuming, and error-prone so I gave up.

File: test/indexer/test_tdr.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+                         uuid=bundle_uuid,
+                         version=one(e['version'] for e in manifest if e['name'] == 'links.json'),
+                         manifest=manifest,
+                         metadata_files=metadata)
+
+    def _make_mock_tdr_tables(self,
+                              source: TDRSource,
+                              bundle_fqid: BundleFQID,
+                              *,
+                              tables: Mapping[str, JSONs] = None) -> None:
```
Comment (hannes-ucsc) [inline test/indexer/test_tdr.py:99]: Optional[…] but see below

File: test/indexer/test_tdr.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        if tables is None:
```
Comment (hannes-ucsc) [inline test/indexer/test_tdr.py:100]: There is only one call site AFAICT, and it doesn't pass `tables` so no need for this conditional or the argument.

File: test/indexer/test_tdr.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+            link['entity']['entity_type'] = 'project'
+            link['entity']['entity_id'] += '_wrong'
+            links_content_column[0] = json.dumps(links_content)
+            with self.assertRaises(RequirementError):
+                self._test_bundle(self.test_source, load_tables=False)
+
+    def _test_bundle(self,
+                     source: TDRSource,
+                     *,
+                     load_tables: bool = True):
```
Comment (hannes-ucsc) [inline test/indexer/test_tdr.py:132]:
```suggestion
                     load_tables: bool):
```

File: test/indexer/test_tdr.py (hunk 7 above)
Comment (hannes-ucsc) [reply test/indexer/test_tdr.py:100]: Look for other cases of unnecessarily complex method/fucntion signatures.

File: test/indexer/test_tdr.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+
+    def test_emulate_bundle(self):
+        with self.subTest('valid links'):
+            self._test_bundle(self.test_source)
+
+        links_content_column = self.tinyquery.tables_by_name[
+            f'{self.test_source.bq_name}.links'
+        ].columns['content'].values
+        links_content = json.loads(one(links_content_column))
+        link = first(link for link in links_content['links']
```
Comment (hannes-ucsc) [inline test/indexer/test_tdr.py:113]: Wrap like this
```
<expr>
for <expr>
if <expr>
```

File: test/indexer/test_tdr.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        self.assertEqual(test_bundle.metadata_files.keys(),
+                         emulated_bundle.metadata_files.keys())
+
+        def key_prefix(key):
+            return key.rsplit('_', 1)[0]
+
+        for key, value in test_bundle.metadata_files.items():
+            # Ordering of entities is non-deterministic so "process_0.json" may
+            # in fact be "process_1.json", etc
+            self.assertEqual(1, len([k
```
Comment (hannes-ucsc) [inline test/indexer/test_tdr.py:148]: Newline after `[`.

IN commit `Wrap at 80 chars`

and check for other cases of this

File: test/indexer/test_tdr.py (hunk 9 above)
Comment (hannes-ucsc) [reply test/indexer/test_tdr.py:113]: IN commit `Wrap at 80 chars`

Comment (hannes-ucsc) [reply test/indexer/test_tdr.py:113]: and check for other cases of this

File: src/azul/json.py [hunk 11]
```diff
@@ -29,6 +31,20 @@ def copy_json(o: JSON) -> MutableJSON:
...
+    """
+    Make a new, mutable copy of a JSON array.
```
Comment (hannes-ucsc) [inline src/azul/json.py:36]:
```suggestion
    Make a new, mutable copy of a JSON array.
    
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/indexer/test_tdr.py [hunk 12]
```diff
@@ -51,80 +42,97 @@
...
     @lru_cache
     def _canned_bundle(self, source: TDRSource) -> TDRBundle:
-        uuid = '1b6d8348-d6e9-406a-aa6a-7ee886e52bf9'
-        version = '2001-01-01T00:00:00.000000Z'
-        path = f'{config.project_root}/test/indexer/data/{uuid}'
-        with open(path + '.metadata.json') as f:
-            metadata = self.convert_metadata(json.load(f))
-        with open(path + '.manifest.json') as f:
-            manifest = self._convert_manifest(json.load(f), metadata, BundleFQID(uuid, version))
+        fqid = BundleFQID(bundle_uuid, '')
```
[the remark this answers is not in the document]
Comment (noah-aviel-dove) [reply test/indexer/test_tdr.py:86]: I don't see any more appropriate value to use. I considered `None` but that violates the type hint. The recanning script never includes the bundle version in the path of the output files, so any value here will have the same effect, i.e. the file won't be found and the `_load_canned_file` will fall back to the un-versioned path.

[the remark this answers is not in the document]
Comment (noah-aviel-dove) [reply test/indexer/test_tdr.py:86]: Actually, I've realized a non-empty fake value would be more appropriate here to assist in debugging anyway. So I'll change to that.

File: test/indexer/test_tdr.py [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+                              bundle_fqid: BundleFQID) -> None:
+        tables = self._load_canned_file(bundle_fqid, 'tables.tdr')['tables']
+        for table_name, table_rows in tables.items():
+            self._make_mock_entity_table(source, table_name, table_rows['rows'])
+
+    def test_emulate_bundle(self):
+        with self.subTest('valid links'):
+            self._test_bundle(self.test_source, load_tables=True)
+
+        links_content_column = self.tinyquery.tables_by_name[
```
Comment (hannes-ucsc) [inline test/indexer/test_tdr.py:103]: Very unwieldy expression. Use local variables to decompose this

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I request the explicit changes below. Additional changes are in the form of TODO comments which can be removed once they are addressed.

```diff
Index: test/indexer/test_tdr.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/indexer/test_tdr.py b/test/indexer/test_tdr.py
--- a/test/indexer/test_tdr.py	(revision 7af7069022492d5a16ff595524f3feafd079e442)
+++ b/test/indexer/test_tdr.py	(date 1611103933214)
@@ -50,7 +50,9 @@
 
 
 class TestTDRPlugin(CannedBundleTestCase):
-    test_source = TDRSource(project='test_project', name='snapshot', is_snapshot=True)
+    test_source = TDRSource(project='test_project',
+                            name='snapshot',
+                            is_snapshot=True)
 
     @cached_property
     def tinyquery(self) -> tinyquery.TinyQuery:
@@ -63,7 +65,9 @@
         self._make_mock_entity_table(source=source,
                                      table_name='links',
                                      rows=[
-                                         dict(links_id=links_id, version=current_version, content='{}')
+                                         dict(links_id=links_id,
+                                              version=current_version,
+                                              content='{}')
                                          for links_id in links_ids
                                      ])
         plugin = TestPlugin(source, self.tinyquery)
@@ -76,16 +80,19 @@
         ])
 
     def _test_source(self, *, is_snapshot: bool) -> TDRSource:
-        return TDRSource(project='foo', name='bar', is_snapshot=is_snapshot)
+        return TDRSource(project='foo',
+                         name='bar',
+                         is_snapshot=is_snapshot)
 
     @lru_cache
     def _canned_bundle(self, source: TDRSource) -> TDRBundle:
         fqid = BundleFQID(bundle_uuid, 'N/A')
         canned_result = self._load_canned_file(fqid, 'result.tdr')
         manifest, metadata = canned_result['manifest'], canned_result['metadata']
+        version = one(e['version'] for e in manifest if e['name'] == 'links.json')
         return TDRBundle(source=source,
                          uuid=bundle_uuid,
-                         version=one(e['version'] for e in manifest if e['name'] == 'links.json'),
+                         version=version,
                          manifest=manifest,
                          metadata_files=metadata)
 
@@ -100,9 +107,9 @@
         with self.subTest('valid links'):
             self._test_bundle(self.test_source, load_tables=True)
 
-        links_content_column = self.tinyquery.tables_by_name[
-            f'{self.test_source.bq_name}.links'
-        ].columns['content'].values
+        dataset = self.test_source.bq_name
+        links_table = self.tinyquery.tables_by_name[dataset + '.links']
+        links_content_column = links_table.columns['content'].values
         links_content = json.loads(one(links_content_column))
         link = first(link
                      for link in links_content['links']
@@ -110,13 +117,21 @@
 
         with self.subTest('invalid supp file entity id'):
             link['entity']['entity_type'] = 'cell_suspension'
+            # TODO: It is not obvious why this is done.
             links_content_column[0] = json.dumps(links_content)
             with self.assertRaises(RequirementError):
+                # TODO: It's not obvious why load_tables=False is passed. That
+                #       should be explained in a comment.
+                # TODO: If I change this to True, this AND the subsequent
+                #       subtest fails. There is no point in marking sections of
+                #       test code as sub tests if they are not indepenedent.
                 self._test_bundle(self.test_source, load_tables=False)
 
         with self.subTest('invalid supp file entity type'):
             link['entity']['entity_type'] = 'project'
             link['entity']['entity_id'] += '_wrong'
+            # TODO: It is not obvious why this is don. If I uncomment this line,
+            #       the test still passes.
             links_content_column[0] = json.dumps(links_content)
             with self.assertRaises(RequirementError):
                 self._test_bundle(self.test_source, load_tables=False)
@@ -140,11 +155,15 @@
         for key, value in test_bundle.metadata_files.items():
             # Ordering of entities is non-deterministic so "process_0.json" may
             # in fact be "process_1.json", etc
+            # TODO: There is a lot of code here to handle the numbering
+            #       non-determinism. I think we should try harder to eliminate it
+            #       instead of accomodating it here.
             self.assertEqual(1, len([
                 k
                 for k, v in emulated_bundle.metadata_files.items()
                 if v == value and key_prefix(key) == key_prefix(k)
             ]))
+
         for manifest in (test_bundle.manifest, emulated_bundle.manifest):
             manifest.sort(key=itemgetter('uuid'))
 
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Note my improvements to the commit titles. All commits addressing an issue should reference the issue in the title. While it's acceptable to have a meandering history in more complex PRs, the meandering part should be clearly grouped such that someone who isn't interested in the changes that are obsoleted by subsequent changes doesn't have to.

When writing tests call the public methods, the ones declared by the super class. Don't call pseudo-public methods that are only public because the test calls them. You'll cover more code and confuse fewer people with surprising test names.

```diff
Index: src/azul/plugins/repository/tdr/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/repository/tdr/__init__.py b/src/azul/plugins/repository/tdr/__init__.py
--- a/src/azul/plugins/repository/tdr/__init__.py	(revision f86f0e0f743dd2b210b5ab59488c728d71728666)
+++ b/src/azul/plugins/repository/tdr/__init__.py	(date 1611870814261)
@@ -112,13 +112,13 @@
 
     def list_bundles(self, prefix: str) -> List[BundleFQID]:
         log.info('Listing bundles in prefix %r.', prefix)
-        bundle_ids = self.list_links_ids(prefix)
+        bundle_ids = self._list_links_ids(prefix)
         log.info('Prefix %r contains %i bundle(s).', prefix, len(bundle_ids))
         return bundle_ids
 
     def fetch_bundle(self, bundle_fqid: BundleFQID) -> Bundle:
         now = time.time()
-        bundle = self.emulate_bundle(bundle_fqid)
+        bundle = self._emulate_bundle(bundle_fqid)
         log.info("It took %.003fs to download bundle %s.%s",
                  time.time() - now, bundle.uuid, bundle.version)
         return bundle
@@ -154,7 +154,7 @@
     def _full_table_name(self, table_name: str) -> str:
         return self._source.qualify_table(table_name)
 
-    def list_links_ids(self, prefix: str) -> List[BundleFQID]:
+    def _list_links_ids(self, prefix: str) -> List[BundleFQID]:
         validate_uuid_prefix(prefix)
         current_bundles = self._query_latest_version(f'''
             SELECT links_id, version
@@ -177,7 +177,7 @@
         else:
             return max(versioned_items, key=itemgetter('version'))
 
-    def emulate_bundle(self, bundle_fqid: BundleFQID) -> Bundle:
+    def _emulate_bundle(self, bundle_fqid: BundleFQID) -> Bundle:
         bundle = TDRBundle(source=self._source,
                            uuid=bundle_fqid.uuid,
                            version=bundle_fqid.version,
Index: test/indexer/test_tdr.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test/indexer/test_tdr.py b/test/indexer/test_tdr.py
--- a/test/indexer/test_tdr.py	(revision f86f0e0f743dd2b210b5ab59488c728d71728666)
+++ b/test/indexer/test_tdr.py	(date 1611871012829)
@@ -57,7 +57,7 @@
     def tinyquery(self) -> tinyquery.TinyQuery:
         return tinyquery.TinyQuery()
 
-    def test_list_links_ids(self):
+    def test_list_bundles(self):
         source = self._test_source(is_snapshot=True)
         current_version = '2001-01-01T00:00:00.000001Z'
         links_ids = ['42-abc', '42-def', '42-ghi', '86-xyz']
@@ -70,7 +70,7 @@
                                          for links_id in links_ids
                                      ])
         plugin = TestPlugin(source, self.tinyquery)
-        bundle_ids = plugin.list_links_ids(prefix='42')
+        bundle_ids = plugin.list_bundles(prefix='42')
         bundle_ids.sort(key=attrgetter('uuid'))
         self.assertEqual(bundle_ids, [
             BundleFQID('42-abc', current_version),
@@ -102,12 +102,12 @@
         for table_name, table_rows in tables.items():
             self._make_mock_entity_table(source, table_name, table_rows['rows'])
 
-    def test_emulate_bundle(self):
+    def test_fetch_bundle(self):
         # Test valid links
-        self._test_bundle(self.test_source, load_tables=True)
+        self._test_fetch_bundle(self.test_source, load_tables=True)
 
-        # Directly modify the contents of the TinyQuery tables to test invalid
-        # links not present in the canned bundle.
+        # Directly modify the canned tables to test invalid links not present
+        # in the canned bundle.
         dataset = self.test_source.bq_name
         links_table = self.tinyquery.tables_by_name[dataset + '.links']
         links_content_column = links_table.columns['content'].values
@@ -115,36 +115,35 @@
         link = first(link
                      for link in links_content['links']
                      if link['link_type'] == 'supplementary_file_link')
-
         # Test invalid entity_type in supplementary_file_link
+        assert link['entity']['entity_type'] == 'project'
         link['entity']['entity_type'] = 'cell_suspension'
         # Update table
         links_content_column[0] = json.dumps(links_content)
-
+        # Invoke code under test
         with self.assertRaises(RequirementError):
-            self._test_bundle(self.test_source,
-                              load_tables=False)  # Avoid resetting tables to canned state
+            self._test_fetch_bundle(self.test_source,
+                                    load_tables=False)  # Avoid resetting tables to canned state
 
         # Undo previous change
         link['entity']['entity_type'] = 'project'
-
         # Test invalid entity_id in supplementary_file_link
         link['entity']['entity_id'] += '_wrong'
         # Update table
         links_content_column[0] = json.dumps(links_content)
-
+        # Invoke code under test
         with self.assertRaises(RequirementError):
-            self._test_bundle(self.test_source, load_tables=False)
+            self._test_fetch_bundle(self.test_source, load_tables=False)
 
-    def _test_bundle(self,
-                     source: TDRSource,
-                     *,
-                     load_tables: bool):
+    def _test_fetch_bundle(self,
+                           source: TDRSource,
+                           *,
+                           load_tables: bool):
         test_bundle = self._canned_bundle(source)
         if load_tables:
             self._make_mock_tdr_tables(source, test_bundle.fquid)
         plugin = TestPlugin(source, self.tinyquery)
-        emulated_bundle = plugin.emulate_bundle(test_bundle.fquid)
+        emulated_bundle = plugin.fetch_bundle(test_bundle.fquid)
 
         self.assertEqual(test_bundle.fquid, emulated_bundle.fquid)
         # Manifest and metadata should both be sorted by entity UUID
```

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
For demo, modify canned bundle and show that test fails. Undo modification and instead modify canned result and show that test fails.

**Pull request merged**